File: typescript-javascript/day04-functions/solutions.js
// Day 4 reference solutions.

function sumAll(...numbers) {
    return numbers.reduce((total, n) => total + n, 0);
}

function greetWithDefault(name, greeting = "Hello") {
    return `${greeting}, ${name}!`;
}

function makeCounter() {
    let count = 0;
    return function () {
        count += 1;
        return count;
    };
}

function makeAdders(n) {
    const indices = [];
    for (let i = 0; i < n; i++) {
        indices.push(i);
    }
    return indices.map((i) => () => i);
}

const bandRoom = {
    name: "The Basement",
    instruments: ["guitar", "drums", "bass"],
    listInstruments: function () {
        return this.instruments.map((instrument) => `${this.name} has a ${instrument}`);
    },
};

module.exports = { sumAll, greetWithDefault, makeCounter, makeAdders, bandRoom };

File: python/day04-functions/solutions.py
"""Day 4 reference solutions."""


def summarize(*numbers, **labels):
    return {"count": len(numbers), "total": sum(numbers), **labels}


def add_item_safely(item, bucket=None):
    if bucket is None:
        bucket = []
    bucket.append(item)
    return bucket


def make_counter():
    count = 0
    def increment():
        nonlocal count
        count += 1
        return count
    return increment


def make_adders(n):
    return [(lambda i=i: i) for i in range(n)]


def apply_to_all(func, items):
    return [func(item) for item in items]


def compose(f, g):
    def h(x):
        return f(g(x))
    return h
